[L2S-471] Documentation of the microservice-package-maven-plugin parameter specification (#3142)
SoftwareAG/c8y-docs: content/microservice-sdk/java-bundle/developing-microservice.md

10% is reserved for "system", but not less than 50 MB. <br>
10% is taken for Metaspace, but not less than 64 MB and not more than 1024MB. <br>
The rest is allocated for heap size.

> Please refer to [package goal](/microservice-sdk/general-aspects/##package-goal) how to change heap and metaspace settings.

Reviewer: ```suggestion
Refer to [Package goal](/microservice-sdk/general-aspects/##package-goal) for information on how to change the heap and metaspace settings.
```

          <version>${c8y.version}</version>
          <executions>
              <execution>
                  <id>package</id>
                  <phase>package</phase>
                  <goals>
                    <goal>package</goal>
                  </goals>
                  <configuration>
                    <name>hello-world</name>
                    <encoding>UTF-8</encoding>
                    <rpmSkip>true</rpmSkip>
                    <containerSkip>false</containerSkip>
                  </configuration>
              </execution>
              <execution>
                  <id>microservice-package</id>
                  <phase>package</phase>
                  <goals>
                    <goal>package</goal>
                  </goals>
                  <configuration>
                    <name>hello-world</name>
                    <image>hello-world</image>
                    <encoding>UTF-8</encoding>
                    <skip>false</skip>
                  </configuration>
              </execution>
          </executions>
      </plugin>
      ...
    </plugins>
    ...
  </build>
  ...
</project>
```

#### Package goal {#package-goal}

> The package plugin is responsible for the creation of a Docker container, rpm file and for creating a ZIP file that can be deployed on the platform.

Reviewer: ```suggestion
The package plugin is responsible for the creation of a Docker container, RPM file and for creating a ZIP file that can be deployed on the platform.
```

It can be configured with the following parameters:

| parameter<br>short form<br>for pom.xml entries<br> in \<configuration> section |                                data type                                |         parmameter command<br>line name          |                                             default value                                              | description                                                                                                                                                                                                                        | 

> |:------------------------------------------------------------------------------:|:-----------------------------------------------------------------------:|:------------------------------------------------:|:------------------------------------------------------------------------------------------------------:|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|  

Reviewer: @jend-c8y An HTML table format would make this table a bit easier to read and to format. It is not necessary to change this now, only a tip for the future.

Author: Ok. Thank you for your hint.

|                                  \<arguments>                                  |                              List\<String>                              |       -Dagent-package.arguments=[...,]...        |                                                   ""                                                   | General command line arguments for jar start up.<br>Specify with "," separated arguments.                                                                                                                                          | 

> |                                \<containerSkip>                                |                                 boolean                                 |        -Dskip.agent.package.container=...        |                                                 false                                                  | Skip the container packaging                                                                                                                                                                                                       |

Reviewer: ```suggestion
|                                \<containerSkip>                                |                                 Boolean                                 |        -Dskip.agent.package.container=...        |                                                 false                                                  | Skip the container packaging                                                                                                                                                                                                       |
```

Author: done

|                                 \<description>                                 |                                 String                                  |            -Dpackage.description=...             |                                         ${project.description}                                         | Microservice description                                                                                                                                                                                                           |  

> |                             \<dockerBuildTimeout>                              |                                   int                                   |  -Dmicroservice.package.dockerBuildTimeout=...   |                                                  360                                                   | Timeout value in seconds for generation of docker build                                                                                                                                                                            | 

Reviewer: ```suggestion
|                             \<dockerBuildTimeout>                              |                                   Int                                   |  -Dmicroservice.package.dockerBuildTimeout=...   |                                                  360                                                   | Timeout value in seconds for generation of docker build                                                                                                                                                                            | 
```

Author: done

|                                  \<encoding>                                   |                                 String                                  |        -Dproject.build.sourceEncoding=...        |                                                 UTF-8                                                  | Define String encoding                                                                                                                                                                                                             |

> |           \<heap><br>\<min>...\</min><br>\<max>...\<max><br>\</heap>           | min, max : \<int>m<br>(for available units refer to java documentation) |                        --                        |                                                   ""                                                   | heap parameter results to -Xms\<min> -Xmx\<max> java runtime arguments for the microservice start up                                                                                                                               |

Reviewer: ```suggestion
|           \<heap><br>\<min>...\</min><br>\<max>...\<max><br>\</heap>           | min, max : \<int>m<br>(for available units refer to Java documentation) |                        --                        |                                                   ""                                                   | heap parameter results to -Xms\<min> -Xmx\<max> Java runtime arguments for the microservice start up                                                                                                                               |
```

Author: done

|                                   \<jvmArgs>                                   |                              List\<String>                              |        -Dagent-package.jvmArgs=[...,]...         | "-XX:+UseG1GC <br>-XX:+UseStringDeduplication <br>-XX:MinHeapFreeRatio=25 <br>-XX:MaxHeapFreeRatio=75" | Java runtime arguments for the microservice start<br>specify with "," separated arguments. Default values will be overwritten if other options are provided. Arguments are passed during microservice startup                      | 
|                                                \<manifestFile>                 |                                 String                                  |                -DmanifestFile=...                |                          "${basedir}/src/main/configuration/cumulocity.json"                           | Path to the microservice manifest file location                                                                                                                                                                                    | 

> |      \<metaspace><br>\<min>...\</min><br>\<max>...\<max><br>\</metaspace>      | min, max : \<int>m<br>(for available units refer to java documentation) |                        --                        |                                                   ""                                                   | \<metaspace> parameter is combined<br>with \<perm> parameter values if available which results in -XX:MetaspaceSize=\<min> -XX:MaxMetaspaceSize=\<max> java runtime arguments for the microservice start up                        |

Reviewer: ```suggestion
|      \<metaspace><br>\<min>...\</min><br>\<max>...\<max><br>\</metaspace>      | min, max : \<int>m<br>(for available units refer to Java documentation) |                        --                        |                                                   ""                                                   | \<metaspace> parameter is combined<br>with \<perm> parameter values if available which results in -XX:MetaspaceSize=\<min> -XX:MaxMetaspaceSize=\<max> Java runtime arguments for the microservice start up                        |
```

Author: done

|                                    \<name>                                     |                                 String                                  |                -Dpackage.name=...                |                                         ${project.artifactId}                                          | Microservice name                                                                                                                                                                                                                  |   

> |           \<perm><br>\<min>...\</min><br>\<max>...\<max><br>\</perm>           | min, max : \<int>m<br>(for available units refer to java documentation) |                        --                        |                                                   ""                                                   | \<perm> parameter is combined with \<metaspace> parameter values for compatibility reasons if available which results in -XX:MetaspaceSize=\<min> -XX:MaxMetaspaceSize=\<max> java runtime arguments for the microservice start up |

Reviewer: ```suggestion
|           \<perm><br>\<min>...\</min><br>\<max>...\<max><br>\</perm>           | min, max : \<int>m<br>(for available units refer to Java documentation) |                        --                        |                                                   ""                                                   | \<perm> parameter is combined with \<metaspace> parameter values for compatibility reasons if available which results in -XX:MetaspaceSize=\<min> -XX:MaxMetaspaceSize=\<max> Java runtime arguments for the microservice start up |
```

Author: done

> |                                   \<rpmSkip>                                   |                                 boolean                                 |           -Dskip.agent.package.rpm=...           |                                                  true                                                  | Skip the rpm packaging                                                                                                                                                                                                             |

Reviewer: ```suggestion
|                                   \<rpmSkip>                                   |                                 Boolean                                 |           -Dskip.agent.package.rpm=...           |                                                  true                                                  | Skip the rpm packaging                                                                                                                                                                                                             |
```

Author: done

> |                                    \<skip>                                     |                                 boolean                                 |             -Dskip.agent.package=...             |                                                 false                                                  | Skip the whole packaging                                                                                                                                                                                                           |

Reviewer: ```suggestion
|                                    \<skip>                                     |                                 Boolean                                 |             -Dskip.agent.package=...             |                                                 false                                                  | Skip the whole packaging                                                                                                                                                                                                           |
```

Author: done

Example configuration in pom.xml:

```xml
...
<plugin>
  <groupId>com.nsn.cumulocity.clients-java</groupId>
  <artifactId>microservice-package-maven-plugin</artifactId>
  <version>${c8y.version}</version>
  <executions>
    <execution>
      <configuration>
        <name>hello-world</name>
        <encoding>UTF-8</encoding>
        <rpmSkip>true</rpmSkip>
        <containerSkip>false</containerSkip>
        <manifestFile>${basedir}/src/main/microservice/cumulocity.json</manifestFile>
        <heap>
          <min>200m</min>
          <max>600m</max>
        </heap>
        <metaspace>
          <min>200m</min>
          <max>300m</max>
        </metaspace>
      </configuration>
    </execution>
  </executions>
</plugin>
...
```

> #### Set heap, perm or metaspace on command line

Reviewer 2: Java PermGen memory space was replaced by Metaspace quite some time ago (Java 8). For the sake of compatibility it is okay that th MS SDK further support configuration related to the "perm" space. However when describing of to Java memory space please do not mention perm space anymore in first place.
```suggestion
#### Set heap or metaspace on command line
```

> Properties must be used if you want to specify data of complex data types on command line. This is the case for memory

Reviewer: ```suggestion
Properties must be used if you want to specify data of complex data types on the command line. This is the case for memory
```

> data like heap, perm and metaspace. In this case you have to specify each primitive value separately as pom property 

Reviewer 2: ```suggestion
data like heap and metaspace. In this case you have to specify each primitive value separately as pom property 
```

which is then used inside the configuration of the microservice-package-maven-plugin. Primitive configuration values 

> or lists/arrays can be set on Maven command line directly without usage of properties, for example ‘-Dskip.agent.package.rpm=”true”’.

Reviewer: ```suggestion
or lists/arrays can be set on the Maven command line directly without usage of properties, for example ‘-Dskip.agent.package.rpm=”true”’.
```

Example pom.xml
```xml
<project … >
  ...
  <properties> 
    ...
    <custom-property.metaspace.min>200m</custom-property.metaspace.min>
    <custom-property.metaspace.max>300m</custom-property.metaspace.max>
    ...
  </properties>
  ...
</project>
```

> These properties now can be used inside the configuration section of the microservice-package-maven-plugin 

Reviewer: ```suggestion
These properties can now be used inside the configuration section of the microservice-package-maven-plugin 
```

Example configuration in pom.xml:

```xml
...
<plugin>
  <groupId>com.nsn.cumulocity.clients-java</groupId>
  <artifactId>microservice-package-maven-plugin</artifactId>
  <version>${c8y.version}</version>
  <executions>
    <execution>
      <configuration>
        <name>hello-world</name>
        <encoding>UTF-8</encoding>
        <rpmSkip>true</rpmSkip>
        <containerSkip>false</containerSkip>
        <manifestFile>${basedir}/src/main/microservice/cumulocity.json</manifestFile>
        <heap>
          <min>200m</min>
          <max>600m</max>
        </heap>
        <metaspace>
          <min>200m</min>
          <max>300m</max>
        </metaspace>
      </configuration>
    </execution>
  </executions>
</plugin>
...
```

#### Set heap or metaspace on command line

> 

Reviewer 3: This chapter contains a mix of pom examples and "command line" (last few lines) -> 2 chapters!?

Author: Adapted the chapter for better understanding.
Created two sub-chapters for command line parameters: primitive and complex ones.
